uc-crux-llvm/src/UCCrux/LLVM/Classify/Types.hs

```diff
@@ -0,0 +1,271 @@
+  | TagDoubleFree
+  deriving (Eq, Ord)
+
+data TruePositive
+  = ConcretelyFailingAssert !What4.ProgramLoc
+  | DoubleFree
+
+truePositiveTag :: TruePositive -> TruePositiveTag
+truePositiveTag =
+  \case
+    ConcretelyFailingAssert {} -> TagConcretelyFailingAssert
+    DoubleFree {} -> TagDoubleFree
+
+ppTruePositiveTag :: TruePositiveTag -> Text
+ppTruePositiveTag =
+  \case
+    TagConcretelyFailingAssert -> "Concretely failing user assertion"
+    TagDoubleFree -> "Double free"
+
+-- | All of the preconditions that we can deduce. We know how to detect and fix
+-- these issues.
+data MissingPreconditionTag
+  = ArgWriteBadAlignment
+  | ArgReadBadAlignment
+  | ArgFreeUnallocated
+  | ArgFreeBadOffset
+  | ArgWriteUnmapped
+  | ArgReadUninitialized
+  | ArgReadUninitializedOffset
+  | ArgPointerConstOffset
+  | ArgMemsetTooSmall
+  | ArgAddSignedWrap
+  | ArgSubSignedWrap
+  deriving (Eq, Ord)
+
+diagnose :: MissingPreconditionTag -> Text
+diagnose =
+  \case
+    ArgWriteBadAlignment -> "Write to a pointer with insufficient alignment in argument"
+    ArgReadBadAlignment -> "Read from a pointer with insufficient alignment in argument"
+    ArgFreeUnallocated -> "`free` called on an unallocated pointer in argument"
+    ArgFreeBadOffset -> "`free` called on pointer with nonzero offset in argument"
+    ArgWriteUnmapped -> "Write to an unmapped pointer in argument"
+    ArgReadUninitialized -> "Read from an uninitialized pointer in argument"
+    ArgReadUninitializedOffset -> "Read from an uninitialized pointer calculated from a pointer in argument"
+    ArgPointerConstOffset -> "Addition of a constant offset to a pointer in argument"
+    ArgMemsetTooSmall -> "`memset` called on pointer to too-small allocation in argument"
+    ArgAddSignedWrap -> "Addition of a constant caused signed wrap of an int in argument"
+    ArgSubSignedWrap -> "Subtraction of a constant caused signed wrap of an int in argument"
+
+prescribe :: MissingPreconditionTag -> Text
+prescribe =
+  ("Prescription: Add a precondition that " <>)
+    . \case
+      ArgWriteBadAlignment -> "the pointer is sufficiently aligned."
+      ArgReadBadAlignment -> "the pointer is sufficiently aligned."
+      ArgFreeUnallocated -> "the pointer points to initialized memory."
+      ArgFreeBadOffset -> "the pointer points to initialized memory."
+      ArgWriteUnmapped -> "the pointer points to allocated memory."
+      ArgReadUninitialized -> "the pointer points to initialized memory."
+      ArgReadUninitializedOffset -> "the pointer points to initialized memory."
+      ArgPointerConstOffset -> "the allocation is at least that size."
+      ArgMemsetTooSmall -> "the allocation is at least that size."
+      ArgAddSignedWrap -> "the integer is small enough"
+      ArgSubSignedWrap -> "the integer is large enough"
+
+-- | There was an error and we know roughly what sort it was, but we still can't
+-- do anything about it.
+data Unfixable
+  = AddSymbolicOffsetToArgPointer
```

Comment:
I'll add a more detailed Haddock to this constructor! In short it really is undefined behavior to even construct a pointer that doesn't point to anything (except one past the end, of course). We can't fix it because it's not clear what kind of precondition we could add to avoid it.

Reply:
I just want to point out that it is undefined behavior to construct (or at least use?) that type of pointer at the source level, but various optimization passes will happily generate program states that do construct that sort of thing (transiently). We might want to be careful of how we apply undefined behavior rules at the level of LLVM bitcode

Reply:
Oh, I see. Well, the memory model definitely generates assertions for this condition, so if we want to allow that we'd have to add a "relaxing" flag like the one for pointer comparisons.

Reply:
I haven't seen this yet, but will adjust as need be.